### DISCUSSION TITLE
How to generate a vsg::scene from points vector?

### DISCUSSION THREAD

**Discussion by WangZP10969** [vsg-dev/VulkanSceneGraph#1063] [Q&A] [answered]
I have tested the vsg example, it works well, loading and displaying model files from local storage. 

If my point cloud is stored in a vector, similar to std::vector<std::array<double,3>>, how to generate a vsg::scene from this vector to display the point cloud, Which example or document should I refer to?

**Comment by robertosfield** [comment]
The vsgPoints library provides support for large point clouds with support for database paging and quantization to enable handling of billion of points in real-time.

This might be overkill for your needs so a custom shader and geometry might be sufficient.  

Rather than a std::array<dobule,3> you can use the vsg::dvec3, but be aware that graphics hardware doesn't process doubles as past as standard floats and not all hardware supports doubles on the GPU.  This means converting to local original and then using vsg::vec3Array would be more efficient.  You can use a vsg::MatrxTransform to place the subgraph that has the local origin into final location in world coords.

**Comment by WangZP10969** [comment]
Thank you very much for your reply! 

I tested the vsg::points example, it’s great, it can read the local mesh and convert it into points for display as a small square.

I didn’t have experience with OSG before, I don’t know how to generate a scene or node using point arrays or triangle arrays. Which example or which class interface documents should I refer to?

**Comment by robertosfield** [comment, marked as answer]
vsgPoints library is for me wondering points. You can specify the points in
a range if ways. Please look at the code in the examples

**Reply by WangZP10969** [reply]
Thank you very much. I found some bricks and createRendering related stuff in the points example. I think this should be useful
    auto settings = vsgPoints::Settings::create();
    settings->createType = vsgPoints::CREATE_FLAT;
    settings->pointSize = 200;
    auto Bricks = vsgPoints::Bricks::create(settings);
    for (double X = -50; X < 50; ++X)
    {
        for (double Y = -50; Y < 50; ++Y)
        {
            Bricks->add({X, Y, 200}, {255, 255, 255, 255});
        }
    }
    auto group = vsg::Group::create();
    if (auto scene = vsgPoints::createSceneGraph(Bricks, Bricks->settings))
    {
        group->addChild(scene);
    }
